Fixed user.present which was breaking when updating workphone,homephone, fullname and "other" fields in case int was passed instead of string

priyank-R opened the pull request:
### What does this PR do?
Fixed user.present which was breaking when updating workphone,homephone, fullname and "other" fields in case int was passed instead of string
### What issues does this PR fix or reference?
Fixes #53961

### Previous Behavior
user.present (for updating the user) was breaking when any of the above-mentioned fields were passed in as an integer (eg 123) and not string (eg: '123') 

### New Behavior
It does not break anymore

### Merge requirements satisfied?
- [x] Docs - Not needed
- [x] Changelog - No sure if its needed for this change.
- [x] Tests written/updated - No.

### Commits signed with GPG?
No

priyank-R committed:
Updated renderer docs mentioning gpg as a text renderer to both, text and data renderer

priyank-R committed:
Merge branch 'master' of https://github.com/saltstack/salt

priyank-R committed:
Merge branch 'master' of https://github.com/saltstack/salt

priyank-R committed:
fixed user.present which was breaking when updating workphone,homephone, fullname, other fields that were passed in as numbers instead of string

priyank-R commented:
@barbaricyawps could you please check this out. Sorry, I'm new here, is it ok to mention you or someone everytime I make a PR ? Thanks

barbaricyawps commented:
@priyank-R , that's totally fine to tag me to review docs-related PRs. I'll make sure I take a look at this later today. Thanks!

priyank-R commented:
@barbaricyawps Thank you kindly.

barbaricyawps commented:
@priyank-R , apologies but since this is not a docs-change I can't accurately review this PR. I am currently only qualified to review docs-related PRs. Currently @MKLeb is assigned to review it. If he's not the appropriate person, he can make sure it gets to the right person for review.

Thanks for your PR and sorry I couldn't be much help!

MKLeb commented:
Hi @priyank-R!  This change makes sense to me, although we do require a changelog and a test before merging.  Would you mind adding those?

priyank-R commented:
Thanks @barbaricyawps. 
@MKLeb yes, I will add the changelog and test case for this. Thanks.

priyank-R committed:
Added file for changelog

MKLeb committed:
Merge branch 'master' into fix_53961

MKLeb commented:
@priyank-R I merged the newer changes from `master` into your branch... all tests now pass.  Just need another small test (unit test most likely) and then this should be good!

priyank-R commented:
Yes @MKLeb , I'm trying to learn a little about how to write Unit tests on salt, and will most likely pull up a small unit test by today. Thanks for all the good help :)

MKLeb commented:
@priyank-R No worries!  Take your time... if you need any additional resources on test writing for salt, let me know!

priyank-R commented:
@MKLeb a silly question but, how do I get something to log in a custom file/console while running states in developement env ? I tried log.error / log.debug if I could get something in /var/log/salt/master, but couldn't. Is there any specific way we can tweak the logging settings in dev?

priyank-R committed:
Merge branch 'master' of https://github.com/saltstack/salt into fix_53961

priyank-R committed:
Added test for gecos_fields in tests/pytests/unit/states/test_user.py

priyank-R committed:
Merge branch 'fix_53961' of github.com:priyank-R/salt into fix_53961

MKLeb commented:
If you want to have the debug output of a salt command displayed to the console, you can do `-l debug` at the end of the command.

For example...
`salt * test.ping -l debug`

priyank-R committed:
Merge branch 'master' of https://github.com/saltstack/salt into fix_53961

priyank-R committed:
Moved the common variables of salt under pytest fixture configure_loader_modules

priyank-R committed:
Removed some comments

priyank-R committed:
Merge branch 'master' into fix_53961

priyank-R committed:
Merge branch 'master' of https://github.com/saltstack/salt into fix_53961

priyank-R committed:
Reverted changes suggested by mkleb in test_user.py

MKLeb commented:
Hi @priyank-R, a few things here...

We have unit tests for this already started in pytest at `tests\pytests\unit\states\test_user.py`.  It already has the loader modules configured for you, so you don't have to do anything besides write a new test in that file!

Here are some resources:
Writing tests for salt: https://docs.saltproject.io/en/latest/topics/development/tests/index.html#writing-tests
Full pytest docs (actually a very easy read): https://docs.pytest.org/en/6.2.x/contents.html
If you like more visual stuff (like myself), a quick youtube tutorial: https://www.youtube.com/watch?v=byaxg00Gf9I

priyank-R commented:
All this while, I was looking into tests/unit directory. Finally found the source and wrote the test. Thank you so very much @MKLeb .

priyank-R commented:
I've added the test for checking the change in gecos fields, please let me know if anything more is to be done. Also, I noticed that I've not been clean in terms of my commits for this particular PR, there were unnecessary mergings and all done by me. That will be taken care by me from next time on. Thank you.

priyank-R commented:
@MKLeb Please check it at your convenience, thanks.

MKLeb commented:
re-run pr-windows-2019-x64-py3-pytest

MKLeb commented on tests/pytests/unit/states/test_user.py:
```diff
@@ -264,3 +264,49 @@ def mock_exists(*args):
+        return_value={"min": 2, "max": 88888, "inact": 77, "warn": 14, "passwd": ""}
+    )
+    shadow_hash = MagicMock(return_value="abcd")
+    dunder_salt = {
+        "user.info": mock_info,
+        "shadow.info": shadow_info,
+        "shadow.default_hash": shadow_hash,
+        "file.group_to_gid": MagicMock(side_effect=["foo"]),
+        "file.gid_to_group": MagicMock(side_effect=[5000, 5000]),
+        "user.chhomephone": MagicMock(return_value=True),
+    }
```
```suggestion
    dunder_salt = {
        "user.info": mock_info,
        "shadow.info": shadow_info,
        "shadow.default_hash": shadow_hash,
        "file.group_to_gid": MagicMock(side_effect=["foo"]),
        "file.gid_to_group": MagicMock(side_effect=[5000, 5000]),
        "user.chhomephone": MagicMock(return_value=True),
    }
```
I believe you can set the `__salt__` variables in the `configure_loader_modules` fixtures.
You can see an example in this file https://github.com/saltstack/salt/blob/master/tests/pytests/unit/modules/test_pip.py

priyank-R replied:
I'll move all the variables except `user.info `and `user.chhomephone` to the fixture, since the 2 I mentioned are specifically for the test I've created, so they won't be of much use outside the test. This is my thinking. I'll make a commit for you to check and you can comment on the same. Thank you

priyank-R replied:
@MKLeb I've made the change as suggested by you. Do let me know if anything is missing.

priyank-R commented:
@MKLeb Please checkout the changes you suggested and let me know if I missed anything. Thanks

MKLeb commented on tests/pytests/unit/states/test_user.py:
```diff
@@ -13,7 +13,18 @@
+        "shadow.default_hash": shadow_hash,
+        "file.group_to_gid": MagicMock(side_effect=["foo"]),
+        "file.gid_to_group": MagicMock(side_effect=[5000, 5000]),
+        # "user.chhomephone": MagicMock(return_value=True),
```
```suggestion
```

MKLeb commented on tests/pytests/unit/states/test_user.py:
```diff
@@ -13,7 +13,18 @@
 @pytest.fixture
 def configure_loader_modules():
-    return {user: {}}
+    shadow_info = MagicMock(
+        return_value={"min": 2, "max": 88888, "inact": 77, "warn": 14, "passwd": ""}
+    )
+    shadow_hash = MagicMock(return_value="abcd")
+    dunder_salt = {
+        "shadow.info": shadow_info,
+        "shadow.default_hash": shadow_hash,
+        "file.group_to_gid": MagicMock(side_effect=["foo"]),
+        "file.gid_to_group": MagicMock(side_effect=[5000, 5000]),
+        # "user.chhomephone": MagicMock(return_value=True),
+    }
+    return {user: {"__salt__": dunder_salt}}
```
```suggestion
    return {user: {}}
```

MKLeb commented on tests/pytests/unit/states/test_user.py:
```diff
@@ -264,3 +275,41 @@ def mock_exists(*args):
+        ]
+    )
+    mock_changes = MagicMock(side_effect=[{"homephone": "667788"}, None])
+    dunder_salt = {
+        "user.info": mock_info,
+        "user.chhomephone": MagicMock(return_value=True),
+    }
```
```suggestion
    shadow_info = MagicMock(
        return_value={"min": 2, "max": 88888, "inact": 77, "warn": 14, "passwd": ""}
    )
    shadow_hash = MagicMock(return_value="abcd")
    dunder_salt = {
        "shadow.info": shadow_info,
        "shadow.default_hash": shadow_hash,
        "file.group_to_gid": MagicMock(side_effect=["foo"]),
        "file.gid_to_group": MagicMock(side_effect=[5000, 5000]),
        "user.info": mock_info,
        "user.chhomephone": MagicMock(return_value=True),
    }
```

MKLeb commented:
Looks like the failing test is unrelated, so I'm good with this.

Also, apologies, but I think I'm going to take back my earlier suggestion.  I think we keep the __salt__ mock local to your test to avoid any unwanted side effects later on.   I have requested it as a change to make it easier for you, but please feel free to test it locally as well.  So sorry about that.

priyank-R commented:
@MKLeb No problem, I've reverted the changes and moved the mocks back to the test. Please let me know if anything more needs to be done. Thanks for your help.

priyank-R commented:
@MKLeb , don't know why tests are failing, perhaps it has nothing to do with my code. But unsure.

MKLeb commented:
@priyank-R So sorry about that, our CI tests are broken at the moment. I'll update your branch once it's done.  I'll review this once the fix is merged and tests are passing.  Thanks for your patience!

priyank-R commented:
@MKLeb sounds good. 
I'm open for more contributions and trying to work on one other issue already. If you have something for me to work on, please feel free to assign to me.

MKLeb committed:
Merge branch 'master' into fix_53961

priyank-R committed:
Merge branch 'master' into fix_53961

priyank-R committed:
Merge branch 'master' into fix_53961

priyank-R committed:
Merge branch 'master' into fix_53961

priyank-R committed:
Merge branch 'master' into fix_53961

priyank-R commented:
@MKLeb looks like all tests are passing now - I've merged the latest changes.

MKLeb commented:
@priyank-R I'm working on getting another set of eyes on this so we can get it merged soon.

twangboy commented:
Looks good to me

waynew commented on tests/pytests/unit/states/test_user.py:
```diff
@@ -264,3 +264,49 @@ def mock_exists(*args):
+    with patch.dict(user.__grains__, {"kernel": "Linux"}), patch.dict(
+        user.__salt__, dunder_salt
+    ), patch.dict(user.__opts__, {"test": False}), patch.object(
+        user, "_changes", mock_changes
```
also FWIW - no need to create a separate mock here. `patch` passes arguments to the underlying Mock, so in this case it could easily be `patch.object(user, "_changes", side_effect=[{"homephone": "667788"}, None])` and it would perform the same.

Makes the test code quite a bit cleaner by eliminating the extra `MagicMock` calls.

priyank-R replied:
Understood @wayne, and you're right, that's a much cleaner way.

waynew reviewed (approved):
Looks test looks good :+1: just a mention about autospec & patch, but nothing to block the PR over.

priyank-R commented on tests/pytests/unit/states/test_user.py:
```diff
@@ -264,3 +264,49 @@ def mock_exists(*args):
+        "shadow.default_hash": shadow_hash,
+        "file.group_to_gid": MagicMock(side_effect=["foo"]),
+        "file.gid_to_group": MagicMock(side_effect=[5000, 5000]),
+    }
```
Wow, never knew about create_autospec method before. Will definitely use it from now on wherever needed, thanks @waynew

priyank-R committed:
Merge branch 'master' into fix_53961

MKLeb committed:
Merge branch 'master' into fix_53961

priyank-R committed:
Merge branch 'master' into fix_53961
